tools/crm_resource_ban.c

```diff
@@ -191,51 +191,130 @@ cli_resource_prefer(const char *rsc_id, const char *host, cib_t * cib_conn)
+    if (rc == -ENXIO) {
+        rc = pcmk_ok;
     }
 
+    free(fragment);
+    return rc;
+}
+
+static int
+resource_clear_node_in_location(const char *rsc_id, const char *host, cib_t * cib_conn)
+{
+    int rc = pcmk_ok;
+    xmlNode *fragment = NULL;
+    xmlNode *location = NULL;
+
     fragment = create_xml_node(NULL, XML_CIB_TAG_CONSTRAINTS);
+    location = create_xml_node(fragment, XML_CONS_TAG_RSC_LOCATION);
+    crm_xml_set_id(location, "cli-ban-%s-on-%s", rsc_id, host);
+
+    location = create_xml_node(fragment, XML_CONS_TAG_RSC_LOCATION);
+    crm_xml_set_id(location, "cli-prefer-%s", rsc_id);
+    if (do_force == FALSE) {
+        crm_xml_add(location, XML_CIB_TAG_NODE, host);
+    }
+
+    crm_log_xml_info(fragment, "Delete");
+    rc = cib_conn->cmds->remove(cib_conn, XML_CIB_TAG_CONSTRAINTS, fragment, cib_options);
+    if (rc == -ENXIO) {
+        rc = pcmk_ok;
+    }
+
+    free(fragment);
+    return rc;
+}
+
+int
+cli_resource_clear(const char *rsc_id, const char *host, GListPtr allnodes, cib_t * cib_conn)
+{
+    int rc = pcmk_ok;
+
+    if(cib_conn == NULL) {
+        return -ENOTCONN;
+    }
 
     if(host) {
-        location = create_xml_node(fragment, XML_CONS_TAG_RSC_LOCATION);
-        crm_xml_set_id(location, "cli-ban-%s-on-%s", rsc_id, host);
+        rc = resource_clear_node_in_expr(rsc_id, host, cib_conn);
+
+        /* rc does not tell us whether the previous operation did anything, only
+         * whether it failed or not.  Thus, as long as it did not fail, we need
+         * to try the second clear method.
+         */
+        if (rc == -ENXIO || rc == pcmk_ok) {
+            rc = resource_clear_node_in_location(rsc_id, host, cib_conn);
+        }
 
     } else {
+        xmlNode *fragment = NULL;
+        xmlNode *location = NULL;
         GListPtr n = allnodes;
+
+        fragment = create_xml_node(NULL, XML_CIB_TAG_CONSTRAINTS);
+
         for(; n; n = n->next) {
             node_t *target = n->data;
 
             location = create_xml_node(fragment, XML_CONS_TAG_RSC_LOCATION);
             crm_xml_set_id(location, "cli-ban-%s-on-%s",
                            rsc_id, target->details->uname);
         }
-    }
 
-    location = create_xml_node(fragment, XML_CONS_TAG_RSC_LOCATION);
-    crm_xml_set_id(location, "cli-prefer-%s", rsc_id);
-    if(host && do_force == FALSE) {
-        crm_xml_add(location, XML_CIB_TAG_NODE, host);
-    }
+        location = create_xml_node(fragment, XML_CONS_TAG_RSC_LOCATION);
+        crm_xml_set_id(location, "cli-prefer-%s", rsc_id);
 
-    crm_log_xml_info(fragment, "Delete");
-    rc = cib_conn->cmds->remove(cib_conn, XML_CIB_TAG_CONSTRAINTS, fragment, cib_options);
-    if (rc == -ENXIO) {
-        rc = pcmk_ok;
+        crm_log_xml_info(fragment, "Delete");
+        rc = cib_conn->cmds->remove(cib_conn, XML_CIB_TAG_CONSTRAINTS, fragment, cib_options);
+        if (rc == -ENXIO) {
+            rc = pcmk_ok;
+        }
```

Comment:
Hmm, I think there are two possible interpretations here. If I'm following correctly, you're thinking that if someone specifies --clear without a host, then they only want to clear ban/move constraints that were placed without specifying a host. I could also see the intent being to clear all ban/move constraints regardless of host. Based on the help message, I think I'd lean to the latter, in which case the no-host path could boil down to a recursive call for each host (or maybe even a single xpath).

Completely unrelated to this PR, the part of the help message "If --node and --force are specified, any constraint created by --move will be cleared, even if it is not for the specified node" makes no sense to me. Being able to clear ban constraints for a specific node only, while clearing all move constraints, seems more confusing than useful to me. I might consider dropping that at some future date.

Reply:
The patch makes it hard to see, but this is just adapted from the existing code.  How I interpreted the existing code was that if no host was specified, it would add a node for every host and the result would be the constraint would be removed from all.  I was hoping to keep that behavior.

So yes, I was going with the second of your interpretations.  I have perhaps not achieved it, though.  I thought it was working in my testing.  I'll try some more things out here.

Reply:
But looking at it some more, I agree.  Making it a recursive call means removing code (always my goal) and reusing what's already there.  It should be a lot clearer, too.